draft-ietf-quic-transport.md

```diff
@@ -3835,108 +3841,136 @@ fields.  The PMTU can depend on path characteristics, and can therefore change
 ICMP message validation MUST include matching IP addresses and UDP ports
 {{!RFC8085}} and, when possible, connection IDs to an active QUIC session.
-
 The endpoint SHOULD ignore all ICMP messages that fail validation.
 
 An endpoint MUST NOT increase PMTU based on ICMP messages; see Section 3, clause
 6 of {{!DPLPMTUD}}.  Any reduction in the QUIC maximum packet size in response
 to ICMP messages MAY be provisional until QUIC's loss detection algorithm
 determines that the quoted packet has actually been lost.
 
+### PMTUD Probes with Handshake packets
 
-## Datagram Packetization Layer PMTU Discovery
-
-Section 6.3 of {{!DPLPMTUD}} provides considerations for implementing Datagram
-Packetization Layer PMTUD (DPLPMTUD) with QUIC.
-
-When implementing the algorithm in Section 5 of {{!DPLPMTUD}}, the initial
-value of BASE_PMTU SHOULD be consistent with the minimum QUIC packet size (1232
-bytes for IPv6 and 1252 bytes for IPv4).
-
-PING and PADDING frames can be used to generate PMTU probe packets. These frames
-might not be retransmitted if a probe packet containing them is lost.  However,
-these frames do consume congestion window, which could delay the transmission of
-subsequent application data.
+One way to construct a PMTU probe is to coalesce (see
+{{packet-coalesce}}) a Handshake packet ({{packet-handshake}}) with a short
+header packet in a single UDP datagram.  If the UDP datagram reaches the
+endpoint, the Handshake packet will be ignored, but the short header packet will
+be acknowledged.  If the UDP datagram causes an ICMP message to be sent, the
+first part of the datagram will be quoted in that message.  If the source
+connection ID is within the quoted portion of the UDP datagram, that could be
+used for routing.
 
-A PING frame can be included in a PMTU probe to ensure that a valid probe is
-acknowledged.
+## Datagram Packetization Layer PMTU Discovery
 
-The considerations for processing ICMP messages in the previous section also
+When implementing the algorithm in Section 5 of {{!DPLPMTUD}}, the
+initial value of BASE_PMTU SHOULD be consistent with the minimum QUIC
+packet size. The MIN_PLPMTU is the same as the BASE_PMTU.
+
+QUIC endpoints implementing DPLPMTUD maintain a maximum packet size 
+(DPLPMTUD MPS) for each combination of local and remote IP
+addresses. 
+If a QUIC endpoint determines that the PLPMTU between any pair of local
+and remote IP addresses has fallen below the size needed to support
+the minimum QUIC packet size (BASE_PLPMTU), it MUST immediately cease
+sending QUIC packets, except for DPLPMTUD probe packets, on the affected
+path. An endpoint MAY terminate the connection if an alternative
+path cannot be found.
+
+###  DPLPMTUD and Initial Connectivity
+
+From the perspective of DPLPMTUD, QUIC transport is an 
+acknowledged PL. A sender can therefore enter the DPLPMTUD BASE
+state when the QUIC connection handshake has been completed and
+the endpoint has established a 1-RTT key.
```

Comment:
This text was proposed by Magnus in a tsvwg thread. We should discuss this.

Reply:
So, this was an attempt from me to correct something that looked even more broken in the earlier text in the TSVWG document. If handshake is complete is a good term lets use it. 

So I think it is relevant to at least go through the implication of sending DPLPMTUD probes prior to handshake complete. So one possible implementation here could be to first send your initial packet with 1280 as MPS. Then you use the rest of your IW=10 to send a broad side of Path MTU probes of various sizes checking a number of common sizes up to the client's local interface's MPS. So in relation to recovery that defines the initial window to 10*max_datagram_size with a limit of max (14720, 2*max_datagram_size). I make an assumption that one will in this case start with max_datagram_size=1200. Thus one can fit some number of probes. So allow DPLPMTUD to use 0-RTT keys to send probes I think is likely to result that the client to server handshake flight will often contain 10-12k of data even if there are no 0-RTT application data. That might be a good investment, but I are we certain of the DoS properties of this? 

So my though was simply to be a bit conservative here and ensure that both endpoints know there are someone there that will report delivery of the probes before sending them. But, this is clearly something the QUIC WG can discuss and change.

Reply:
@gorryfair replies (below, repeated to keep the thread intact):

> The reason why we didn't suggest this within DPLPMTUD, was two things:
>
> * Using the limited for IW for packet probes rapidly consumes the current IW space in bytes, and we were unconvinced of the importance to tenatively try a large PMTU during the first RTT.
> 
> * I was not aware of measurement data about how changing the initial flight of data would impact congestion collapse, since at this point there is no feedback from the receiver, and it seems, at least to me, to raise questions regarding stability with large numbers of flows and also as you note, the potential DoS opportunities.
>
> So, I'd be interested in what people think about this, and see any analysis and experience.

Reply:
I think that the intent of the text is fine, I was only concerned about the execution, purely from a logical standpoint.  That is, `(handshake complete) && (got 1-RTT keys) == (handshake complete)` because `(got 1-RTT keys)` is a precondition for `(handshake complete)`.

As you both say, there are concerns with probing before this point.  Maybe you can, and indeed it is probably beneficial.  But maybe it is also risky, and we don't need that.  If anyone does want to probe before this point, they are off the books for now.  If anyone wants to explicitly allow that in the draft, that's a design change that needs support, justification, and a separate issue to track.  I don't think we can do that as part of this change.

Reply:
This seems like the best direction to me - I leave to you whether (got 1-RTT keys) needs to be mentioned as not, and included that in your PR in case you wish to edit (or not).